jfix-stdlib-ratelimiter/src/main/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcher.kt

```diff
@@ -0,0 +1,282 @@
+                name, state.get()
+            )
+            return
+        }
+        rateLimiter.close()
+        profiler.detachIndicator(QUEUE_SIZE_INDICATOR)
+        profiler.detachIndicator(ACTIVE_ASYNC_OPERATIONS)
+    }
+
+
+    private suspend fun <T> processTask(task: Task2<T>): T {
+        task.queueWaitTimeCall.stop()
+        try {
+            if (windowSize > 0) {
+                profiler.start("acquire_window").use { acquireWindowTime ->
+                    var windowAcquired = false
+                    while (!windowAcquired) {
+                        if (state.get() == State.TERMINATE) {
+                            val message = "RateLimitedDispatcher [$name] received TERMINATE state while window acquiring"
+                            logger.error(message)
+                            throw RejectedExecutionException(message)
+                        }
+                        windowAcquired = semaphore.get().tryAcquire()
+                        if (!windowAcquired) {
+                            delay(WINDOW_ACQUIRE_DELAY_MS)
+                        }
+                    }
+                    acquireWindowTime.stop()
+                }
+            }
+            profiler.start("acquire_limit").use { limitAcquireTime ->
+                var limitAcquired = false
+                while (!limitAcquired) {
+                    if (state.get() == State.TERMINATE) {
+                        val message = "RateLimitedDispatcher [$name] received TERMINATE state while rate limiter acquiring"
+                        logger.error(message)
+                        throw RejectedExecutionException(message)
+                    }
+                    limitAcquired = rateLimiter.tryAcquire(3, ChronoUnit.SECONDS)
+                }
+                limitAcquireTime.stop()
+            }
+
+            queueSize.decrementAndGet()
+            // Since async operation may complete faster then Started method call
+            // it must be called before asynchronous operation started
+            asyncOperationStarted()
+
+            return profiler.profile(
+                    "supply_operation",
+                    Supplier { task.supplier.invoke() }
+            )!!
+        } catch (e: Exception) {
+            logger.error("RateLimitedDispatcher [$name] received exception: $e")
+            throw RejectedExecutionException("RateLimitedDispatcher [$name] received exception: ${e.message}")
+        }
+    }
+
+    private open class Task2<T>(
+            val supplier: () -> T,
+            val queueWaitTimeCall: ProfiledCall
+    )
+
+    private interface Command {
+        suspend fun apply()
+    }
+
+    private inner class ChangeWindowSizeCommand(private val oldSize: Int, private val newSize: Int) :
+        Command {
+        override suspend fun apply() {
+            if (newSize == oldSize) return
+            semaphore.getAndUpdate {
```

Comment:
From `getAndUpdate` docs:
> The function should be side-effect-free, since it may be re-applied when attempted updates fail due to contention among threads.

Currently, it is not sife-effect free. 

Actually, there are a lot of concurrency problems since you calling this command in some unknown DynamicProperty listener thread. Compared to old RateLimitedDispatcher, where all operations were handled by one specific thread. 

You have 2 options:

1. fix all concurrency problems with window size changing command applying. To help you - you might consider restriction of some critical section of code to be executed only by one coroutine at time (see coroutines Mutex)
2. (Preferred by me) To do not implement window size completely. Historically, it was implemented for jfix-armeria-limiter to limit number of concurrently running futures. But back pressure is required to be implemented in other place. Thus, there is no big need in window size support (at least currently).

Reply:
Checked, window uses only in ru.fix.cpapsm.brt.proxy.discovery.config.CommonBrtProxyDiscoveryConfig.Companion#createRateLimitedDispatcher.

Window (and semaphor) logic removed